stdlib/public/runtime/ImageInspectionELF.cpp

```diff
@@ -35,67 +37,124 @@ static const char ProtocolConformancesSymbol[] =
 static const char TypeMetadataRecordsSymbol[] =
   ".swift2_type_metadata_start";
 
+static std::atomic<bool> didInitializeProtocolConformanceLookup;
+static std::atomic<bool> didInitializeTypeMetadataLookup;
+
 /// Context arguments passed down from dl_iterate_phdr to its callback.
 struct InspectArgs {
   /// Symbol name to look up.
   const char *symbolName;
   /// Callback function to invoke with the metadata block.
   void (*addBlock)(const void *start, uintptr_t size);
+  /// Set to true when initialize*Lookup() is called.
+  std::atomic<bool> *didInitializeLookup;
 };
 
-static int iteratePHDRCallback(struct dl_phdr_info *info,
-                               size_t size, void *data) {
-  const InspectArgs *inspectArgs = reinterpret_cast<const InspectArgs *>(data);
-  void *handle;
-  if (!info->dlpi_name || info->dlpi_name[0] == '\0') {
-    handle = dlopen(nullptr, RTLD_LAZY);
-  } else {
-    handle = dlopen(info->dlpi_name, RTLD_LAZY | RTLD_NOLOAD);
-  }
+static InspectArgs ProtocolConformanceArgs = {
+  ProtocolConformancesSymbol,
+  addImageProtocolConformanceBlockCallback,
+  &didInitializeProtocolConformanceLookup
+};
 
-  if (!handle) {
-    // Not a shared library.
-    return 0;
-  }
+static InspectArgs TypeMetadataRecordArgs = {
+  TypeMetadataRecordsSymbol,
+  addImageTypeMetadataRecordBlockCallback,
+  &didInitializeTypeMetadataLookup
+};
 
-  const char *conformances =
-    reinterpret_cast<const char*>(dlsym(handle, inspectArgs->symbolName));
 
-  if (!conformances) {
-    // if there are no conformances, don't hold this handle open.
+// Extract the section information for a named section in an image. imageName
+// can be nullptr to specify the main executable.
+static SectionInfo getSectionInfo(const char *imageName,
+                                  const char *sectionName) {
+  SectionInfo sectionInfo = { 0, nullptr };
+  void *handle = dlopen(imageName, RTLD_LAZY);
+  if (handle) {
+    void *symbol = dlsym(handle, sectionName);
+    if (symbol) {
+      // Extract the size of the section data from the head of the section.
+      const char *section = reinterpret_cast<const char *>(symbol);
+      memcpy(&sectionInfo.size, section, sizeof(uint64_t));
+      sectionInfo.data = section + sizeof(uint64_t);
+    }
     dlclose(handle);
+  }
+  return sectionInfo;
+}
+
+static int iteratePHDRCallback(struct dl_phdr_info *info,
+                               size_t size, void *data) {
+  const InspectArgs *inspectArgs = reinterpret_cast<const InspectArgs *>(data);
+  const char *fname = info->dlpi_name;
+
+  // While dl_iterate_phdr() is in progress it holds a lock to prevent other
+  // images being loaded. The initialize flag is set here inside the callback so
+  // that addNewDSOImage() sees a consistent state. If it was set outside the
+  // dl_interate_phdr() call then it could result in images being missed or
+  // added twice.
+  inspectArgs->didInitializeLookup->store(true, std::memory_order_release);
```

Comment:
If it's guaranteed that `dlopen` and `dl_iterate_phdr` hold the same lock, then the `didInitializeLookup` flag should be able to be nonatomic.